### PR TITLE
add tagalign linter

### PR DESCRIPTION
**Pull request by 4meepo**
[tagalign](https://github.com/4meepo/tagalign) is used to align and sort tags in golang struct. 

eg:
```go
type FooBar struct {
    Foo    int    `json:"foo" validate:"required"`
    Bar    string `json:"bar" validate:"required"`
    FooFoo int8   `json:"foo_foo" validate:"required"`
}
```
can be aligned to 
```go
type FooBar struct {
    Foo    int    `json:"foo"     validate:"required"`
    Bar    string `json:"bar"     validate:"required"`
    FooFoo int8   `json:"foo_foo" validate:"required"`
}
```

It can also sort tags with user's prefered order.

### CODE REVIEW

**Comment by ldez** [comment]
In order for a pull request adding a linter to be reviewed, the linter and the PR must follow some requirements.

### Pull Request Description

- [ ] It must have a link to the linter repository.
- [ ] It must provide a short description of the linter.

### Linter

- [ ] It must not be a duplicate of another linter or a rule of a linter. (the team will help to verify that)
- [ ] It must have a valid license (AGPL is not allowed) and the file must contain the required information by the license, ex: author, year, etc.
- [ ] The linter repository must have a CI and tests.
- [ ] It must use [`go/analysis`](https://golangci-lint.run/contributing/new-linters/).
- [ ] It must have a valid tag, ex: `v1.0.0`, `v0.1.0`.
- [ ] It must not contain `init()`.
- [ ] It must not contain `panic()`, `log.fatal()`, `os.exit()`, or similar.
- [ ] It must not have false positives/negatives. (the team will help to verify that)
- [ ] It must have tests inside golangci-lint.

### The Linter Tests Inside Golangci-lint

- [ ] They must have at least one std lib import.
- [ ] They must work with `T=<name of the linter test file>.go make test_linters`. (the team will help to verify that)

### `.golangci.reference.yml`

- [ ] The linter must be added to the list of available linters (alphabetical case-insensitive order).
    - `enable` and `disable` options
- [ ] If the linter has a configuration, the exhaustive configuration of the linter must be added (alphabetical case-insensitive order)
    - The values must be different from the default ones.
    - The default values must be defined in a comment.
    - The option must have a short description.

### Others Requirements

- [ ] The files (tests and linter) inside golangci-lint must have the same name as the linter.
- [ ] The `.golangci.yml` of golangci-lint itself must not be edited and the linter must not be added to this file.
- [ ] The linters must be sorted in the alphabetical order (case-insensitive) in the `Manager.GetAllSupportedLinterConfigs(...)` and `.golangci.reference.yml`.
- [ ] The load mode (`WithLoadMode(...)`):
    - if the linter doesn't use types: `goanalysis.LoadModeSyntax`
    - `goanalysis.LoadModeTypesInfo` required `WithLoadForGoAnalysis()` in the `Manager.GetAllSupportedLinterConfigs(...)`
- [ ] The version in `WithSince(...)` must be the next minor version (`v1.X.0`) of golangci-lint.

### Recommendations

- [ ] The linter should not use SSA. (currently, SSA does not support generics)
- [ ] The linter repository should have a readme and linting.
- [ ] The linter should be published as a binary. (useful to diagnose bug origins)

---

The golangci-lint team will edit this comment to check the boxes before and during the review.

If the author of the PR is a member of the golangci-lint team, he should not edit this message.

**This checklist does not imply that we will accept the linter.**

**Comment by 4meepo** [comment]
hi team, ready for review.

File: .golangci.reference.yml [hunk 1]
```diff
@@ -1708,6 +1708,16 @@ linters-settings:
     # Default: ["200", "400", "404", "500"]
     http-status-code-whitelist: [ "200", "400", "404", "500" ]
 
+  tagalign:
+    # whether enable tags auto sort. if true, the tags will be sorted by name in ascending order.
+    # eg: `xml:"bar" json:"bar" validate:"required"` -> `json:"bar" validate:"required" xml:"bar"`
+    # Default: false
+    auto-sort: true
+    # spcify the fixed order of tags, the other tags will be sorted by name.
```
Comment (alexandear) [inline .golangci.reference.yml:1716]:
```suggestion
    # Specify the fixed order of tags, the other tags will be sorted by name.
```

File: .golangci.reference.yml [hunk 2]
```diff
@@ -1708,6 +1708,16 @@ linters-settings:
     # Default: ["200", "400", "404", "500"]
     http-status-code-whitelist: [ "200", "400", "404", "500" ]
 
+  tagalign:
+    # whether enable tags auto sort. if true, the tags will be sorted by name in ascending order.
```
Comment (alexandear) [inline .golangci.reference.yml:1712]:
```suggestion
    # Whether enable tags auto sort. If true, the tags will be sorted by name in ascending order.
```

File: .golangci.reference.yml [hunk 3]
```diff
@@ -1708,6 +1708,16 @@ linters-settings:
...
+    # eg: `xml:"bar" json:"bar" validate:"required"` -> `json:"bar" validate:"required" xml:"bar"`
+    # Default: false
+    auto-sort: true
+    # spcify the fixed order of tags, the other tags will be sorted by name.
+    # this options will be ignored if `auto-sort` is false.
```
Comment (alexandear) [inline .golangci.reference.yml:1717]:
```suggestion
    # This option will be ignored if `auto-sort` is false.
```

File: test/testdata/configs/tagalign.yml [hunk 4]
```diff
@@ -0,0 +1,4 @@
+linters-settings:
+  tagalign:
+    auto-sort: true
```
Comment (alexandear) [inline test/testdata/configs/tagalign.yml:3]: Could we add a test when `auto-sort` is `false`?

File: pkg/golinters/tagalign.go [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+					options = append(options, tagalign.WithAutoSort(settings.FixedOrder...))
+				} else {
+					options = append(options, tagalign.WithAutoSort())
+				}
+			}
+
+			tagalignIssues := tagalign.Run(p, options...)
+
+			issues := make([]goanalysis.Issue, len(tagalignIssues))
+			for idx, issus := range tagalignIssues {
```
Comment (alexandear) [inline pkg/golinters/tagalign.go:39]:
```suggestion
			for i, issues := range tagalignIssues {
```

File: test/testdata/configs/tagalign.yml (hunk 4 above)
Comment (4meepo) [reply test/testdata/configs/tagalign.yml:3]: Done. Can be tested by 
```bash
T=tagalign_no_sort.go make test_linters
```

**Comment by 4meepo** [comment]
@alexandear  Really thanks for your suggestion. A new test case with auto-sort disabled have been added. 

Try:
```bash
T=tagalign_no_sort.go make test_linters
```

**Comment by ldez** [comment]
The modification require by the checklist is not done.

@alexandear It's better to wait before reviewing that the user part of the checklist is done.

**Comment by 4meepo** [comment]
If there are anything I need to do, just let me know. : )

**Comment by ldez** [comment]
Follow the checklist expect boxes with the message "(the team will help to verify that)"

**Comment by ldez** [comment]
IMO the alignment of the tags is less interesting than the sort of the tags, maybe it can be a good idea to add an option to enable/disable the alignment.
Without this option, I think I will never use this linter.

But this improvement is not a blocker to merging this PR.

**Comment by 4meepo** [comment]
> IMO the alignment of the tags is less interesting than the sort of the tags, maybe it can be a good idea to add an option to enable/disable the alignment. Without this option, I think I will never use this linter.
> 
> But this improvement is not a blocker to merging this PR.

Nice suggestion. I have added a `align` configuration option, which is enabled by default. Now the alignment and sort can be used together or separately.

**Comment by ldez** [comment]
I change a bit the default, the new default: `align:true` and `sort:true`
Because I think your linter will be used more if the sort is enabled by default.

Otherwise, I modified the tests to have simpler test cases, and I see that some options don't work as expected: if `align:false` the option `sort` and `order` don't work.

**Comment by 4meepo** [comment]
Found a bug [here](https://github.com/4meepo/tagalign/commit/d8a15bcced9e93e748363eb63177a187379bf8cf).

Should work now.

**Review by ldez — APPROVED** [review]
Thanks for your contribution :+1:

**Pull request merged**